### DISCUSSION TITLE
bootlogo xavier bmp partition size exceeded

### DISCUSSION THREAD

**Discussion by maggu2810** [OE4T/meta-tegra#1425] [Q&A] [answered]
I am using a custom distribution for my Xavier NX.
The distribution is using mender.

Now I would like to use a custom bootlogo.
I create a bbappend as described here: https://github.com/OE4T/meta-tegra/discussions/670

```
IMAGE_NAME = "test.png"

PR .= ".1"
FILESEXTRAPATHS:prepend := "${THISDIR}/files:"
SRC_URI:append = " \
                  file://${IMAGE_NAME} \
                 "
PACKAGECONFIG = "raw"
RAW_GRAPHIC_FILE = "${IMAGE_NAME}"
SPLASH_GRAPHIC_BACKGROUND = "black"
BMP_BLOB_SIZES = "800x480"

do_configure() {
    mv "${WORKDIR}/${IMAGE_NAME}" "${S}/${RAW_GRAPHIC_FILE}"
}
```

After building the image I would like to update using mender as usual.
The update failed:

```
INFO[0075] Collected output (stderr) while running script /var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update
Error: BUP contents too large for boot partition
ERR: bootloader update failed

---------- end of script output 
INFO[0075] Output (stdout) from command "/var/lib/mender/scripts/ArtifactInstall_Leave_80_bl-update":   Processing BMP_b... [FAIL] 
```

Hm, okay, I added a logo so perhaps the partition layout has been changed.

I tried a `./doflash.sh`

Also this command failed with

```
[ 131.6604 ] Writing partition BMP with bmp.blob
[ 132.9181 ] 0000000074740204: E> NV3P_SERVER: Accessing offset 201178 after boundary partition size 196608
```

I assume I need to modify the flash.xml and increase the size for BMP and BMP_b. Is this correct?

bmp.blob is current 201178 bytes.
flash.xml defines a size of 196608 for the bmp partitions.

If changing the partition layout is the way to go, would you use a factor of the current logo size if it every increases (to be prepared)? How much size do you expect for an image? Will this be constant if I use one defined width x height?

* I created a `mender-custom-flash-layout_%.bbappend`
* copied `repos/meta-mender-community/meta-mender-tegra/recipes-bsp/tegra-binaries/mender-custom-flash-layout/jetson-xavier-nx-devkit-emmc/flash_mender.xml` (as this is my base machine)
* modified the sizeof BMP and BMP_b partition in the copy of `flash_mender.xml` (ATM I doubled the size from 196608 to 393216).

But now I get an error while building the image:

```
...
| [   0.9409 ] tegrahost_v2 --chip 0x19 --align mem_rcm.bct
| [   0.9413 ]
| [   0.9429 ] tegrahost_v2 --chip 0x19 0 --magicid MEMB --ratchet_blob ratchet_blob.bin --appendsigheader mem_rcm.bct zerosbk
| [   0.9432 ] adding BCH for mem_rcm.bct
| [   0.9440 ]
| [   0.9462 ] tegrasign_v3.py --key None --list mem_rcm_sigheader.bct_list.xml --pubkeyhash pub_key.key
| [   0.9463 ] Assuming zero filled SBK key
| [   0.9483 ] tegrahost_v2 --chip 0x19 0 --updatesigheader mem_rcm_sigheader.bct.encrypt mem_rcm_sigheader.bct.hash zerosbk
| [   0.9490 ]
| [   0.9490 ] Copying signatures
| [   0.9505 ] tegrahost_v2 --chip 0x19 0 --partitionlayout flash.xml.bin --updatesig images_list_signed.xml
| [   1.2681 ]
| [   1.2697 ] tegraparser_v2 --generategpt --pt flash.xml.bin
| [   1.2700 ] End sector for bpmp-fw-dtb_b, expected at: 65502, actual: 65535
| [   1.2701 ]
| Error: Return value 4
| Command tegraparser_v2 --generategpt --pt flash.xml.bin
| WARNING: exit code 1 from a shell command.
ERROR: Task (/yocto.src/layers/meta-tegra/recipes-kernel/kernel-bup-payload/kernel-bup-payload.bb:do_deploy) failed with exit code '1'
```

"end sector" ... There is just the size defined for each partition. What am I doing wrong?

**Comment by madisongh** [comment, marked as answer]
Bear in mind that the SPI flash is only 16MiB, and is already pretty full with the default layout. Doubling the size of your BMP partitions has caused you to run out of space, which that error message is rather obliquely telling you.

It would be better to prioritize keeping your logo size within the already-defined limit.  If you really can't do that, you could increase the partition size by 24KiB or so (each) without running into problems, I think.  Anything more than that, and you should carefully map what's where in the flash (including what might be used by `tegra-boot-tools` for your Mender setup, if this is an SDcard-based Xavier) to make sure you don't run out of space and that there's no overlapping use by different parts of the system.

**Reply by maggu2810** [reply]
Thank you very much, I could have guessed that.

The message sounded to me as if the end of a partition was not cleanly aligned and not that the total size of the flash memory had been exceeded.
I also assumed that if I only use one resolution and "only" 800x480, that I would be far below the actual size limit. But since the whole thing is lz4 compressed, it probably always depends on how well the bitmap can be compressed.

I can't find any really helpful information about another "problem". Can I just ask the question here too?
The resolution of the connected display I use is 800x480. My probably naive assumption was that if I set BMP_BLOB_SIZES to 800x480 that I would then see the full image on the display.
However, it is rather the case that a small part of the actual image is still visible at the bottom right and otherwise only black.
Do I have to patch the resolution in the bootloader (cboot)? I suspect:
* cboot-t19x/32.7.4-r0/cboot-t19x-32.7.4/bootloader/partner/common/drivers/display/edid/tegrabl_edid.c
* cboot-t19x/32.7.4-r0/cboot-t19x-32.7.4/bootloader/partner/common/drivers/display/edid/tegrabl_modes.c
* cboot-t19x/32.7.4-r0/cboot-t19x-32.7.4/bootloader/partner/common/drivers/display/edid/tegrabl_modes.h

In addition, why can I specify more resolutions here in BMP_BLOB_SIZES (or pass them to the NVIDIA tooling behind it)?

**Reply by madisongh** [reply]
Multiple bitmaps can be included to handle monitors of different sizes.  If you have just one monitor you're ever going to use, you should be able to include just the one bitmap.

And yeah, if cboot doesn't already have a mode entry for that resolution, you probably need to patch it to add one.
